File: web/lib/api-client.ts
import { RunningConfig, LogCallback } from './utils';

export class SportsUploaderError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SportsUploaderError';
  }
}

export async function makeRequest(
  method: 'GET' | 'POST',
  url: string,
  headers: Record<string, string>,
  params?: Record<string, string>,
  data?: string,
  logCb?: LogCallback
): Promise<any> {
  try {
    const timeoutValue = 15000;
    
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeoutValue);
    
    let requestUrl = url;
    if (method === 'GET' && params) {
      const searchParams = new URLSearchParams(params);
      requestUrl += '?' + searchParams.toString();
    }
    
    const requestOptions: RequestInit = {
      method,
      headers,
      signal: controller.signal,
    };
    
    if (method === 'POST' && data) {
      requestOptions.body = data;
    }
    
    const response = await fetch(requestUrl, requestOptions);
    clearTimeout(timeoutId);
    
    if (!response.ok) {
      const errorText = await response.text();
      logCb?.(`HTTP error occurred: ${response.status} ${response.statusText}`, 'error');
      logCb?.(`URL: ${url}`, 'error');
      logCb?.(`Response text: ${errorText}`, 'error');
      throw new SportsUploaderError(`HTTP Error: ${response.status} ${response.statusText}`);
    }
    
    const result = await response.json();
    return result;
  } catch (error) {
    if (error instanceof SportsUploaderError) {
      throw error;
    }
    
    if (error instanceof Error) {
      if (error.name === 'AbortError') {
        logCb?.('Request timeout occurred', 'error');
        throw new SportsUploaderError('Request timeout');
      }
      
      logCb?.(`An unexpected error occurred: ${error.message}`, 'error');
      throw new SportsUploaderError(`Request Error: ${error.message}`);
    }
    
    throw new SportsUploaderError('Unknown error occurred');
  }
}

export async function getAuthorizationTokenAndRules(
  config: RunningConfig,
  logCb?: LogCallback
): Promise<[string, any]> {
  const commonAppHeaders = {
    "Host": config.HOST,
    "Connection": "keep-alive",
    "sec-ch-ua-platform": "\"Android\"",
    "User-Agent": "Mozilla/5.0 (Linux; Android 12; SM-S9080 Build/V417IR; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/138.0.7204.67 Safari/537.36 TaskCenterApp/3.5.0",
    "Accept": "application/json, text/plain, */*",
    "sec-ch-ua": "\"Not)A;Brand\";v=\"8\", \"Chromium\";v=\"138\", \"Android WebView\";v=\"138\"",
    "Content-Type": "application/json;charset=utf-8",
    "sec-ch-ua-mobile": "?0",
    "X-Requested-With": "edu.sjtu.infoplus.taskcenter",
    "Sec-Fetch-Site": "same-origin",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Dest": "empty",
    "Referer": "https://pe.sjtu.edu.cn/phone/",
    "Accept-Encoding": "gzip, deflate, br, zstd",
    "Accept-Language": "zh-CN,zh;q=0.9,en-US;q=0.8,en;q=0.7",
    "Cookie": config.COOKIE
  };

  logCb?.(`Attempting to get Authorization Token from: ${config.UID_URL}`);
  const uidResponseData = await makeRequest('GET', config.UID_URL, commonAppHeaders, undefined, undefined, logCb);

  let authToken: string;
  if (uidResponseData?.code === 0 && uidResponseData?.data?.uid) {
    authToken = uidResponseData.data.uid;
    logCb?.(`Successfully retrieved Authorization Token: ${authToken}`);
  } else {
    throw new SportsUploaderError(`Failed to get Authorization Token: ${JSON.stringify(uidResponseData)}`);
  }

  logCb?.(`Attempting to get MyData from: ${config.MY_DATA_URL}`);
  try {
    await makeRequest('GET', config.MY_DATA_URL, commonAppHeaders, undefined, undefined, logCb);
    logCb?.('Successfully sent MyData request.');
  } catch (error) {
    logCb?.(`Warning: Failed to get MyData (this might be expected or ignorable): ${error}`, 'warning');
  }

  const formattedLon = config.START_LONGITUDE.toFixed(14);
  const formattedLat = config.START_LATITUDE.toFixed(14);
  const currentLocationParam = `${formattedLon},${formattedLat}`;

  const refererUrlForPointRule = `${config.POINT_RULE_URL}?location=${encodeURIComponent(currentLocationParam)}`;

  const pointRuleHeaders = {
    "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
    "accept-language": "zh-CN,zh;q=0.9,en;q=0.8",
    "authorization": authToken,
    "cache-control": "no-cache",
    "pragma": "no-cache",
    "sec-ch-ua": "\"Chromium\";v=\"140\", \"Not=A?Brand\";v=\"24\", \"Google Chrome\";v=\"140\"",
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": "\"Windows\"",
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-origin",
    "upgrade-insecure-requests": "1",
    "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36",
    "referer": refererUrlForPointRule,
    "Host": config.HOST,
  };

  const params = { location: currentLocationParam };
  
  logCb?.(`Getting point rules from: ${config.POINT_RULE_URL} with location: ${currentLocationParam}`);
  const pointRuleResponseData = await makeRequest('GET', config.POINT_RULE_URL, pointRuleHeaders, params, undefined, logCb);

  return [authToken, pointRuleResponseData?.data || {}];
}

export async function uploadRunningData(
  config: RunningConfig,
  authToken: string,
  runningData: any[],
  logCb?: LogCallback
): Promise<any> {
  const headers = {
    "Authorization": authToken,
    "Content-Type": "application/json; charset=utf-8",
    "Host": config.HOST,
    "Connection": "Keep-Alive",
    "Accept-Encoding": "gzip",
    "User-Agent": "okhttp/4.10.0"
  };

  const response = await makeRequest(
    'POST',
    config.UPLOAD_URL,
    headers,
    undefined,
    JSON.stringify(runningData),
    logCb
  );
  
  logCb?.(`Upload Response: ${JSON.stringify(response, null, 2)}`);
  return response;
}

File: src/api_client.py
import requests
import json
from urllib.parse import quote
from src.utils import log_output, SportsUploaderError

def make_request(method, url, headers, params=None, data=None, log_cb=None, stop_check_cb=None):
    """通用HTTP请求函数"""
    try:
        if stop_check_cb and stop_check_cb():
            log_output("API请求被中断。", "warning", log_cb)
            raise SportsUploaderError("任务已停止。")

        timeout_value = 15

        response = None

        if method.upper() == 'GET':
            response = requests.get(url, headers=headers, params=params, timeout=timeout_value)
        elif method.upper() == 'POST':
            response = requests.post(url, headers=headers, data=data, timeout=timeout_value)
        else:
            raise ValueError(f"Unsupported HTTP method: {method}")

        if stop_check_cb and stop_check_cb():
            log_output("API响应已获取，但任务被中断。", "warning", log_cb)
            raise SportsUploaderError("任务已停止。")

        response.raise_for_status()
        return response.json()
    except requests.exceptions.HTTPError as e:
        log_output(f"HTTP error occurred: {e}", "error", log_cb)
        log_output(f"URL: {url}", "error", log_cb)
        if response is not None:
            log_output(f"Response status code: {response.status_code}", "error", log_cb)
            log_output(f"Response text: {response.text}", "error", log_cb)
            try:
                log_output(f"Response JSON (if any): {json.dumps(response.json(), indent=2)}", "error", log_cb)
            except json.JSONDecodeError:
                log_output(f"Response text (non-JSON): {response.text}", "error", log_cb)
        raise SportsUploaderError(f"HTTP Error: {e}")
    except requests.exceptions.ConnectionError as e:
        log_output(f"Connection error occurred: {e}", "error", log_cb)
        if isinstance(e.args[0], requests.packages.urllib3.exceptions.MaxRetryError) and e.args[0].reason:
            log_output(f"Underlying reason: {e.args[0].reason}", "error", log_cb)
        raise SportsUploaderError(f"Connection Error: {e}")
    except requests.exceptions.Timeout as e:
        log_output(f"Timeout error occurred: {e}", "error", log_cb)
        raise SportsUploaderError(f"Timeout Error: {e}")
    except requests.exceptions.RequestException as e:
        log_output(f"An unexpected error occurred: {e}", "error", log_cb)
        raise SportsUploaderError(f"Request Error: {e}")
    except json.JSONDecodeError:
        log_output(f"Failed to decode JSON from response: {response.text if response else 'No response'}", "error", log_cb)
        raise SportsUploaderError(f"JSON Decode Error: {response.text if response else 'No response'}")


def get_authorization_token_and_rules(config, log_cb=None, stop_check_cb=None):
    """
    通过GET请求获取Authorization Token，并随后获取跑步规则。
    """
    common_app_headers = {
        "Host": config["HOST"],
        "Connection": "keep-alive",
        "sec-ch-ua-platform": "\"Android\"",
        "User-Agent": "Mozilla/5.0 (Linux; Android 12; SM-S9080 Build/V417IR; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/138.0.7204.67 Safari/537.36 TaskCenterApp/3.5.0",
        "Accept": "application/json, text/plain, */*",
        "sec-ch-ua": "\"Not)A;Brand\";v=\"8\", \"Chromium\";v=\"138\", \"Android WebView\";v=\"138\"",
        "Content-Type": "application/json;charset=utf-8",
        "sec-ch-ua-mobile": "?0",
        "X-Requested-With": "edu.sjtu.infoplus.taskcenter",
        "Sec-Fetch-Site": "same-origin",
        "Sec-Fetch-Mode": "cors",
        "Sec-Fetch-Dest": "empty",
        "Referer": "https://pe.sjtu.edu.cn/phone/",
        "Accept-Encoding": "gzip, deflate, br, zstd",
        "Accept-Language": "zh-CN,zh;q=0.9,en-US;q=0.8,en;q=0.7",
        "Cookie": config["COOKIE"]
    }

    log_output(f"Attempting to get Authorization Token from: {config['UID_URL']}", callback=log_cb)
    uid_response_data = make_request('GET', config['UID_URL'], common_app_headers, log_cb=log_cb, stop_check_cb=stop_check_cb)

    auth_token = None
    if uid_response_data.get('code') == 0 and 'uid' in uid_response_data.get('data', {}):
        auth_token = uid_response_data['data']['uid']
        log_output(f"Successfully retrieved Authorization Token: {auth_token}", callback=log_cb)
    else:
        raise SportsUploaderError(f"Failed to get Authorization Token: {uid_response_data}")

    if stop_check_cb and stop_check_cb():
        log_output("任务被请求停止，正在退出...", "warning", log_cb)
        raise SportsUploaderError("任务已停止。")

    log_output(f"\nAttempting to get MyData from: {config['MY_DATA_URL']}", callback=log_cb)
    try:
        make_request('GET', config['MY_DATA_URL'], common_app_headers, log_cb=log_cb, stop_check_cb=stop_check_cb)
        log_output(f"Successfully sent MyData request.", callback=log_cb)
    except Exception as e:
        log_output(f"Warning: Failed to get MyData (this might be expected or ignorable): {e}", "warning", log_cb)

    if stop_check_cb and stop_check_cb():
        log_output("任务被请求停止，正在退出...", "warning", log_cb)
        raise SportsUploaderError("任务已停止。")

    formatted_lon_for_param = f"{config['START_LONGITUDE']:.14f}"
    formatted_lat_for_param = f"{config['START_LATITUDE']:.14f}"
    current_location_param = f"{formatted_lon_for_param},{formatted_lat_for_param}"

    referer_url_for_point_rule = f"{config['POINT_RULE_URL']}?location={quote(current_location_param, safe='')}"

    point_rule_headers = {
        "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7",
        "accept-language": "zh-CN,zh;q=0.9,en;q=0.8",
        "authorization": auth_token,
        "cache-control": "no-cache",
        "pragma": "no-cache",
        "sec-ch-ua": "\"Chromium\";v=\"140\", \"Not=A?Brand\";v=\"24\", \"Google Chrome\";v=\"140\"",
        "sec-ch-ua-mobile": "?0",
        "sec-ch-ua-platform": "\"Windows\"",
        "sec-fetch-dest": "empty",
        "sec-fetch-mode": "cors",
        "sec-fetch-site": "same-origin",
        "upgrade-insecure-requests": "1",
        "user-agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36",
        "referer": referer_url_for_point_rule,
        "Host": config["HOST"],
    }

    params_string = f"?location={quote(current_location_param, safe='')}"
    url = config["POINT_RULE_URL"]

    log_output(f"\nGetting point rules from: {url} with location: {current_location_param}", callback=log_cb)
    point_rule_response_data = make_request('GET', url + params_string, point_rule_headers, log_cb=log_cb, stop_check_cb=stop_check_cb)

    return auth_token, point_rule_response_data.get('data', {})

def upload_running_data(config, auth_token, running_data, log_cb=None, stop_check_cb=None):
    """
    上传跑步数据到服务器。
    """
    headers = {
        "Authorization": auth_token,
        "Content-Type": "application/json; charset=utf-8",
        "Host": config["HOST"],
        "Connection": "Keep-Alive",
        "Accept-Encoding": "gzip",
        "User-Agent": "okhttp/4.10.0"
    }

    if stop_check_cb and stop_check_cb():
        log_output("任务被请求停止，正在退出...", "warning", log_cb)
        raise SportsUploaderError("任务已停止。")

    response = make_request(
        'POST',
        config["UPLOAD_URL"],
        headers,
        data=json.dumps(running_data),
        log_cb=log_cb,
        stop_check_cb=stop_check_cb
    )
    log_output(f"Upload Response: {json.dumps(response, indent=2)}", callback=log_cb)
    return response